src/d3a/models/strategy/external_strategies/storage.py

```diff
@@ -0,0 +1,261 @@
+                 "error_message": f"Error when handling bid delete "
+                                  f"on area {self.device.name} with arguments {arguments}."})
+
+    def _bid(self, payload):
+        bid_response_channel = f'{self.device.name}/bid/response'
+        if not check_for_connected_and_reply(self.redis, bid_response_channel, self.connected):
+            return
+        try:
+            arguments = json.loads(payload["data"])
+            assert set(arguments.keys()) == {'price', 'energy'}
+            arguments['buyer_origin'] = self.device.name
+        except Exception:
+            self.redis.publish_json(
+                bid_response_channel,
+                {"error": "Incorrect bid request. Available parameters: (price, energy)."}
+            )
+        else:
+            self.pending_requests.append(
+                IncomingRequest("bid", arguments, bid_response_channel))
+
+    def _bid_impl(self, arguments, bid_response_channel):
+        try:
+            self.state.clamp_energy_to_buy_kWh([self.market.time_slot])
+            max_energy = self.state.energy_to_buy_dict[self.market.time_slot]
+            assert arguments["energy"] <= max_energy
+
+            bid = self.post_bid(
+                self.market,
+                arguments["price"],
+                arguments["energy"],
+                buyer_origin=arguments["buyer_origin"]
+            )
+            self.state.offered_buy_kWh[self.market.time_slot] += arguments["energy"]
+            self.redis.publish_json(bid_response_channel,
+                                    {"status": "ready", "bid": bid.to_JSON_string()})
+        except Exception as e:
+            logging.error(f"Error when handling bid create on area {self.device.name}: "
+                          f"Exception: {str(e)}, Bid Arguments: {arguments}")
+            self.redis.publish_json(
+                bid_response_channel,
+                {"status": "error",
+                 "error_message": f"Error when handling bid create "
+                                  f"on area {self.device.name} with arguments {arguments}."})
+
+    def event_market_cycle(self):
+        self.register_on_market_cycle()
+        if self.connected:
+            self.state.market_cycle(self.market_area.current_market.time_slot,
+                                    self.market.time_slot)
+            self.state.clamp_energy_to_sell_kWh([self.market.time_slot])
+            market_event_channel = f"{self.device.name}/market_event"
+            current_market_info = self.market.info
+            current_market_info["free_storage"] = self.state.free_storage(self.market.time_slot)
+            current_market_info["used_storage"] = self.state.used_storage
+            current_market_info["min_allowed_soc_ratio"] = self.state.min_allowed_soc_ratio
+            current_market_info["capacity"] = self.state.capacity
+            current_market_info["max_abs_battery_power_kW"] = self.state.max_abs_battery_power_kW
```

Comment:
These values should be known beforehand right? IMO we shouldnt clamp our message queue with this redundant stuff.

Reply:
In addition to that, there are 2 stats that would be very useful and are currently missing, the amount of available energy to buy and sell in this market slot. I would recommend to add these 2 values and remove min_allowed_soc_ratio, capacity, max_abs_battery_power_kW